src/v/archival/archival_metadata_stm.cc

```diff
@@ -456,19 +458,26 @@ command_batch_builder& command_batch_builder::update_highest_producer_id(
 
 ss::future<std::error_code> command_batch_builder::replicate() {
     _as.check();
-    return _stm.get()._lock.with([this]() {
-        vlog(
-          _stm.get()._logger.debug, "command_batch_builder::replicate called");
-        auto now = ss::lowres_clock::now();
-        auto timeout = now < _deadline ? _deadline - now : 0ms;
-        return _stm.get().do_sync(timeout, &_as).then([this](bool success) {
-            if (!success) {
-                return ss::make_ready_future<std::error_code>(errc::not_leader);
-            }
-            auto batch = std::move(_builder).build();
-            return _stm.get().do_replicate_commands(std::move(batch), _as);
-        });
-    });
+
+    auto units = co_await _stm.get()._lock.get_units(_as);
+
+    vlog(_stm.get()._logger.debug, "command_batch_builder::replicate called");
+    auto now = ss::lowres_clock::now();
+    auto timeout = now < _deadline ? _deadline - now : 0ms;
+
+    // Block on syncing the STM.
+    auto did_sync = co_await _stm.get().do_sync(timeout, &_as);
+    if (!did_sync) {
+        co_return errc::not_leader;
+    }
+
+    auto batch = std::move(_builder).build();
+    auto f = _stm.get()
+               .do_replicate_commands(std::move(batch), _as)
+               .finally([u = std::move(units)] {});
```

Comment:
I think that the gate should be held here as well.

Reply:
Done